addon/components/polaris-dropzone.js

```diff
@@ -0,0 +1,522 @@
+   * () => void
+   *
+   * @type {Function}
+   * @default no-op
+   * @public
+   */
+  onDragEnter() {},
+
+  /**
+   * Callback triggered when one or more files left the drag area
+   * () => void
+   *
+   * @type {Function}
+   * @default no-op
+   * @public
+   */
+  onDragLeave() {},
+
+  /**
+   * Callback triggered when one or more files are dragging over the drag area
+   * () => void
+   *
+   * @type {Function}
+   * @default no-op
+   * @public
+   */
+  onDragOver() {},
+
+  /**
+   * Callback triggered on any file drop
+   * (files, acceptedFiles, rejectedFiles) => void
+   *
+   * @type {Function}
+   * @default no-op
+   * @public
+   */
+  onDrop() {},
+
+  /**
+   * Callback triggered when at least one of the files dropped was accepted
+   * (acceptedFiles) => void
+   *
+   * @type {Function}
+   * @default no-op
+   * @public
+   */
+  onDropAccepted() {},
+
+  /**
+   * Callback triggered when at least one of the files dropped was rejected
+   * (rejectedFiles) => void
+   *
+   * @type {Function}
+   * @default no-op
+   * @public
+   */
+  onDropRejected() {},
+
+  /**
+   * Callback triggered when the file dialog is canceled
+   * () => void
+   *
+   * @type {Function}
+   * @default no-op
+   * @public
+   */
+  onFileDialogClose() {},
+
+  iconDragDrop,
+  iconAlertCircle,
+
+  state: computed(() => State.create()).readOnly(),
+
+  isDragging: or('active', 'state.dragging').readOnly(),
+
+  fileInputNode: computed(function() {
+    return this.$(`#${ this.get('elementId') }-input:first`);
+  }).readOnly(),
+
+  ariaDisabled: computed('disabled', function() {
+    if (isPresent(this.get('disabled'))) {
+      return 'true';
+    }
+
+    return null;
+  }).readOnly(),
+
+  sizeClass: computed('state.size', function() {
+    let size = this.get('state.size');
+    return `Polaris-DropZone--size${ capitalize(size) }`;
+  }).readOnly(),
+
+  showDragOverlay: computed('isDragging', 'state.error', 'overlay', function() {
+    let error = this.get('state.error');
+    let { isDragging, overlay } = this.getProperties('isDragging', 'overlay');
+
+    return isDragging && !error && overlay;
+  }).readOnly(),
+
+  /**
+   * Event handlers
+   */
+  handleClick(event) {
+    let { onClick, disabled } = this.getProperties('onClick', 'disabled');
+    if (disabled) {
+      return;
+    }
+
+    return onClick ? onClick(event) : this.open();
+  },
+
+  handleDrop(event) {
+    event.preventDefault();
+    event.stopPropagation();
+
+    let { disabled, onDrop, onDropAccepted, onDropRejected } = this.getProperties(
+      'disabled',
+      'onDrop',
+      'onDropAccepted',
+      'onDropRejected',
+    );
+    if (disabled) {
+      return;
+    }
+    let fileList = getDataTransferFiles(event);
+    let { files, acceptedFiles, rejectedFiles } = this.getValidatedFiles(fileList);
+
+    this.dragTargets = [];
+
+    this.get('state').setProperties({
+      dragging: false,
+      error: rejectedFiles.length > 0,
```

Comment:
As things stand if `rejectedFiles` is somehow `null`/`undefined` etc., this will blow up. `isPresent` will effectively check for both `null`/`undefined` _and_ populated. But I understand wanting to keep this close to the React implementation for easier maintenance 👍

Reply:
`rejectedFiles` should never be anything except an array, otherwise it's a bug in the code